internal/credit/balance_connector.go

```diff
@@ -0,0 +1,428 @@
+package credit
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"slices"
+	"time"
+
+	"github.com/openmeterio/openmeter/internal/streaming"
+	"github.com/openmeterio/openmeter/pkg/framework/entutils"
+	"github.com/openmeterio/openmeter/pkg/models"
+)
+
+// Generic connector for balance related operations.
+type BalanceConnector interface {
+	GetBalanceOfOwner(ctx context.Context, owner NamespacedGrantOwner, at time.Time) (*GrantBalanceSnapshot, error)
+	GetBalanceHistoryOfOwner(ctx context.Context, owner NamespacedGrantOwner, params BalanceHistoryParams) (GrantBurnDownHistory, error)
+	ResetUsageForOwner(ctx context.Context, owner NamespacedGrantOwner, at time.Time) (balanceAfterReset *GrantBalanceSnapshot, err error)
+}
+
+type BalanceHistoryParams struct {
+	From time.Time
+	To   time.Time
+}
+
+func NewBalanceConnector(
+	gc GrantDBConnector,
+	bsc BalanceSnapshotDBConnector,
+	oc OwnerConnector,
+	sc streaming.Connector,
+	log *slog.Logger,
+) BalanceConnector {
+	return &balanceConnector{
+		gc:     gc,
+		bsc:    bsc,
+		oc:     oc,
+		sc:     sc,
+		logger: log,
+
+		// TODO: make configurable
+		snapshotGracePeriod: time.Hour,
+	}
+}
+
+type balanceConnector struct {
+	// grants and balance snapshots are managed in this same package
+	gc  GrantDBConnector
+	bsc BalanceSnapshotDBConnector
+	// external dependencies
+	oc     OwnerConnector
+	sc     streaming.Connector
+	logger *slog.Logger
+
+	snapshotGracePeriod time.Duration
+}
+
+var _ BalanceConnector = &balanceConnector{}
+
+func (m *balanceConnector) GetBalanceOfOwner(ctx context.Context, owner NamespacedGrantOwner, at time.Time) (*GrantBalanceSnapshot, error) {
+	// get last valid grantbalances
+	balance, err := m.getLastValidBalanceSnapshotForOwnerAt(ctx, owner, at)
+	if err != nil {
+		return nil, err
+	}
+
+	periodStart, err := m.oc.GetUsagePeriodStartAt(ctx, owner, at)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current usage period start for owner %s at %s: %w", owner.ID, at, err)
+	}
+	if balance.At.Before(periodStart) {
+		// This is an inconsistency check. It can only happen if we lost our snapshot for the last reset.
+		//
+		// The engine doesn't manage rollovers at usage reset so it cannot be used to calculate GrantBurnDown across resets.
+		return nil, fmt.Errorf("last valid balance snapshot %s is before current period start at %s, no snapshot was created for reset", balance.At, periodStart)
+	}
+
+	// get all relevant grants
+	grants, err := m.gc.ListActiveGrantsBetween(ctx, owner, balance.At, at)
```

Comment:
you mean this part?
```go
// get all relevant grants
grants, err := m.gc.ListActiveGrantsBetween(ctx, owner, period.From, period.To)
// these grants might not be present in the starting balance so lets fill them
for _, grant := range grants {
	if _, ok := balance.Balances[grant.ID]; !ok {
		if grant.ActiveAt(period.From) {
			// This is only possible in case the grant becomes active exactly at the start of the current period
			balance.Balances.Set(grant.ID, grant.Amount)
		} else {
			balance.Balances.Set(grant.ID, 0.0)
		}
	}
}
```

Reply:
yes

Reply:
deduped, see elsewhere